src/timely-util/src/builder_async.rs

```diff
@@ -0,0 +1,344 @@
+    /// *inputs*, that any timestamp appearing at the input, any output timestamp will be greater
+    /// than or equal to the input timestamp subjected to a `Summary` greater or equal to some
+    /// element of the corresponding antichain in `connection`.
+    ///
+    /// Commonly the connections are either the unit summary, indicating the same timestamp might
+    /// be produced as output, or an empty antichain indicating that there is no connection from
+    /// the input to the output.
+    pub fn new_output_connection<D: Container>(
+        &mut self,
+        connection: Vec<Antichain<<G::Timestamp as Timestamp>::Summary>>,
+    ) -> (
+        OutputWrapper<G::Timestamp, D, TeeCore<G::Timestamp, D>>,
+        StreamCore<G, D>,
+    ) {
+        self.builder.new_output_connection(connection)
+    }
+
+    /// Creates an operator implementation from supplied logic constructor.
+    pub fn build<B, L>(self, constructor: B)
+    where
+        B: FnOnce(Vec<Capability<G::Timestamp>>) -> L,
+        L: Future + 'static,
+    {
+        let operator_waker = self.operator_waker;
+        let registered_wakers = self.registered_wakers;
+        let shared_frontiers = self.shared_frontiers;
+        self.builder.build_reschedule(move |caps| {
+            let mut logic_fut = Some(Box::pin(constructor(caps)));
+            move |new_frontiers| {
+                // First, discover if there are any frontier notifications
+                {
+                    let mut old_frontiers = shared_frontiers.borrow_mut();
+                    for ((old, pending), new) in old_frontiers.iter_mut().zip(new_frontiers) {
+                        if PartialOrder::less_than(&old.borrow(), &new.frontier()) {
+                            *old = new.frontier().to_owned();
+                            *pending = true;
+                        }
+                    }
+                }
+                // Then, wake up any registered Wakers for the input streams while taking care to
+                // not reactivate the timely operator since that would lead to an infinite loop.
+                // This ensures that the input handles wake up properly when managed by other
+                // executors, e.g a `select!` that provides its own Waker implementation.
+                {
+                    operator_waker.active.store(true, Ordering::SeqCst);
+                    let mut registered_wakers = registered_wakers.borrow_mut();
+                    for waker in registered_wakers.drain(..) {
+                        waker.wake();
+                    }
+                    operator_waker.active.store(false, Ordering::SeqCst);
+                }
+                // Schedule the logic future if any of the wakers above marked the task as ready
+                if let Some(fut) = logic_fut.as_mut() {
+                    if operator_waker.task_ready.load(Ordering::SeqCst) {
```

Comment:
let me make sure I understand this:
`registered_wakers` are woken on every timely activations, to unblock futures managed by some other executor, but we only ever poll our future if we wake a `TimelyWaker` right? So an unblocked executor could make progress, but until it produces a result that wakes the core `logic_fut`, we do not poll the logic future?

Reply:
That is exactly correct, yeah